### DISCUSSION TITLE
Nginx Reverse Proxy Breaks RAG

### DISCUSSION THREAD

**Discussion by http403** [open-webui/open-webui#3111] [Q&A]
I have put a minimal reverse proxy in front, and for some reason breaks web socket, and makes web search not functional.

default.conf
```
server {
    listen 80 default_server;
    server_name _;

    location / {
        proxy_pass http://ui:8080;

        include proxy.conf;
    }

    location /ollama/api {
        proxy_pass http://ollama:11434/api;

        include proxy.conf;
    }
}
```

proxy.conf
```
proxy_set_header Host $host;
proxy_set_header X-Real-IP $remote_addr;
proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
proxy_set_header X-Forwarded-Proto $scheme;
```
Success screenshot
![A screenshot of Open-Web UI with successful web search](https://github.com/open-webui/open-webui/assets/8369976/9fb989e2-23e4-4093-ac97-4e1e10201231)

Failed screenshot
![A screenshot of Open-Web UI with failed web search](https://github.com/open-webui/open-webui/assets/8369976/3b2d7f01-6c79-4fbf-ad7d-7ca6c179ea95)

Most websocket connection are successful but only fail (seemingly) when doing a web search. I also attached two logs file, one showing [success](https://github.com/user-attachments/files/15811952/rag-web.log), and the other [failed](https://github.com/user-attachments/files/15811953/rag-web-fail.log) from the UI. Hopes they help.

**UPDATE**: Here a log of [nginx](https://github.com/user-attachments/files/15812051/proxy-rag-web-ws-fail.log), and [ui](https://github.com/user-attachments/files/15812052/rag-web-ws-fail.log) when the browser attempts to connect the web socket but failed. Interestingly, nginx, ui, and Firefox all shows a 200 status code yet failed/refused to connect.
![A screenshot of the Firefox Dev Tool showing failed web socket connection](https://github.com/open-webui/open-webui/assets/8369976/eced83c2-1069-42fd-8315-639f1a423edf)

The path is `/ws/socket.io/?EIO=4&transport=websocket&sid=2hOuveEpPBZ42em3AAAI`

**UPDATE 2**: It seems the RAG is totally botched. Uploading to document tab will cause the LM ignore the document. Uploading in chat will cause the whole thing hangs. It must be something seriously wrong with my Nginx setup.

**Comment by spammenotinoz** [comment]
For Nginx to support web-sockets you need to add these lines inyour location block
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";

I am not using Nginx with this project, but likely you also need something like
client_max_body_size 100M; added in your http section, so uploads greater up to 100M will be allowed.  I believe 1mb is the default limit, so anything larger will fail.

**Reply by http403** [reply]
Added. Doesn't do anything.

Web search isn't included to the context so the response is the "I don't know" type. Chat with images completely hangs on the frontend where the CPU completely returns back to idle. I didn't test with the document but I suspect the same.

**Comment by rvkwi** [comment]
Not sure if it helps you, but i had some issues on the last or so update that my https broke (needed for the mic support on mobile). was also related to websockets i believe, had llama adjust my proxy a bit and everything works for me again (including RAG).

So, can't exactly tell you which part might be missing on yours but maybe it helps to reference against:

```
$ cat nginx/nginx.conf 
server {
    listen 443 ssl;
    server_name hts;

    ssl_certificate /etc/ssl/ssl.crt;
    ssl_certificate_key /etc/ssl/ssl.key;

    location / {
        proxy_pass http://hts:3000/;
        proxy_http_version 1.1;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;

        # WebSocket support
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";

        proxy_buffering off;

        # Ensure sub_filter is in use to adjust paths
        sub_filter 'src="/' 'src="/open-webui/';
        sub_filter 'href="/' 'href="/open-webui/';
        sub_filter_once off;
    }

    # Handle static files directly if necessary
    location /open-webui/ {
        alias /path/to/static/files/;  
        try_files $uri $uri/ =404;
    }
}

```

**Reply by http403** [reply]
Nope. This is what I got now
```
server {
    listen 80 default_server;
    server_name _;

    location / {
        proxy_pass http://ui:8080;

        # include proxy.conf;
        proxy_http_version 1.1;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $http_x_forwarded_for,$remote_addr;
        proxy_set_header X-Forwarded-Proto $scheme;

        # WebSocket support
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";

        proxy_buffering off;

        client_max_body_size 100M;
    }

    location /ollama/ {
        rewrite ^/ollama(/.*)$ $1 break;

        proxy_pass http://ollama:11434;

        proxy_http_version 1.1;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $http_x_forwarded_for,$remote_addr;
        proxy_set_header X-Forwarded-Proto $scheme;

        # WebSocket support
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";

        proxy_buffering off;

        # Timeout
        proxy_read_timeout 600s;
        client_body_timeout 300s;
        keepalive_timeout 300s;

        client_max_body_size 100M;
    }
}
```
And URL inserted via `#https://example.com` still not working. It seems the content isn't inserted into the prompt for the LLM to reference.
